src/proxy-animation.js

```diff
@@ -30,151 +30,470 @@ function createReadyPromise(details) {
-        playState: "idle",
-        readyPromise: null,
-        sequence: 0, /* Used to track ready promises. */
-        aborted: new Set(), /* Aborted sequences. */
-      });
-      return;
-    }
-    const isScrollAnimation = timeline && timeline instanceof ScrollTimeline;
+    const animation =
+        (effect instanceof nativeAnimation) ?
+           effect : new nativeAnimation(effect, animationTimeline);
+    const isScrollAnimation = timeline instanceof ScrollTimeline;
     const animationTimeline = isScrollAnimation ? undefined : timeline;
     proxyAnimations.set(this, {
-      animation: new nativeAnimation(effect, animationTimeline),
+      animation: animation,
       timeline: isScrollAnimation ? timeline : undefined,
       playState: isScrollAnimation ? "idle" : null,
       readyPromise: null,
+      // Start and hold times are directly tracked in the proxy despite being
+      // accessible via the animation so that direct manipulation of these
+      // properties does not affect the play state of the underlying animation.
+      // Note that any changes to these values require an update of current
+      // time for the underlying animation to ensure that its hold time is set
+      // to the correct position.
+      startTime: null,
+      holdTime: null,
+      // When changing the timeline on a paused animation, we defer updating the
+      // start time until the animation resumes playing.
+      resetCurrentTimeOnResume: false,
+      // Calls to reverse and updatePlaybackRate set a pending rate that does
+      // not immediately take effect. The value of this property is
+      // inaccessible via the web animations API and therefore explicitly
+      // tracked.
+      pendingPlaybackRate: null,
+      proxy: this,
       sequence: 0, /* Used to track ready promises. */
       aborted: new Set(), /* Aborted sequences. */
     });
   }
 
+  // -----------------------------------------
+  // Helper method
+  // -----------------------------------------
+
+  tick(timelineTime) {
+    if (timelineTime != null && this.playState == 'running') {
+      proxyAnimations.get(this).animation.currentTime =
+          (timelineTime - this.startTime) * this.playbackRate;
+      updateFinishedState(proxyAnimations.get(this));
+    }
+  }
+
+  // -----------------------------------------
+  // Web animation API
+  // -----------------------------------------
+
   get effect() {
     return proxyAnimations.get(this).animation.effect;
   }
   set effect(newEffect) {
-    return proxyAnimations.get(this).animation.effect = newEffect;
+    proxyAnimations.get(this).animation.effect = newEffect;
   }
 
   get timeline() {
-    const internalTimeline = proxyAnimations.get(this).timeline;
-    if (internalTimeline !== undefined)
-      return internalTimeline;
-    return proxyAnimations.get(this).animation.timeline;
+    const details = proxyAnimations.get(this);
+    return details.timeline || details.animation.timeline;
```

Comment:
Comment that if we set a null timeline we revert to the underlying animation's timeline.